### ISSUE TITLE
Searching for literal * keys return all when * unavailable

### ISSUE THREAD

**Issue by micimize** [s3u/JSONPath#77]
Searching for a literal `*` with `$['*']` yields the following unexpected behavior:
```javascript
JSONPath({json: {'*': 'star', 'rest': 'rest', 'foo': 'bar'}, path: "$['*']"})
//=> [ 'star' ]
JSONPath({json: {'rest': 'rest', 'foo': 'bar'}, path: "$['*']"})
//=> [ 'rest', 'bar' ]
```
It's possible to work around this with `$[?(@property === '*')]`, but this isn't usable in more complex queries, such as `$[?(@property === '*'), rest]`

**Comment by micimize** [comment]
Actually this also applies in reverse:
```javascript
JSONPath({json: {'*': 'star', 'rest': 'rest', 'foo': 'bar'}, path: "$.*"})
//=> [ 'star' ]
```

**Comment by brettz9** [comment]
I've added a commit to `master` which adds the breaking change to always treat special characters as such unless a backtick (\`) is added at the beginning in which case it will always treat as a literal.

If this will end up as a problem, I can revert, but I think this behavior makes the most sense and is the safest.

**Comment by micimize** [comment]
Thanks for the quick response! This works in the flat structures, but leads to some unexpected results with more complex nested structures:
```javascript
let json = {'*': 'star', 'foo': 'bar', 'sibling': { child: 'sibling child', '*': 'sibling star'}}

// earlier tests work as expected
JSONPath({ json, path: "$[foo,`*]" })
// => [ 'bar', 'star' ]

// sibling subqueries ignore `*
JSONPath({ json, path: "$[`*,sibling[child,`*]]" })
// => [ 'sibling child', 'sibling star' ]

// ordering of subqueries queries changes results
JSONPath({ json, path: "$[sibling[child,`*],`*]" })
// => [ 'sibling child', 'sibling star', 'sibling star' ]
```
I'm also curious as to why you chose ``` ` ``` instead of `\`, which seems more intuitive to me

**Comment by brettz9** [comment]
The current parsing is regex based. I think we'd need to move to something more robust to deal properly with the nesting. I also don't have a lot of time to work on this project atm. Feel free to submit any PRs to handle though (along with your test cases).

As far as the backtick, I just used that (along the lines of templates, etc.) because I hadn't provided support for escaping of individual characters but rather for the whole sequence.

**Comment by micimize** [comment]
ah, I see - the `,` operator only works in the tail position of a path, making that issue more related to #66  
That's what I thought with the syntax decision - makes sense 👍

**Comment by brettz9** [comment]
FYI, the feature is included with v0.16.0, now released.
